Fix the issue where the prices for most Alibaba Cloud instance types are incorrect when the instance family generation number is greater than or equal to 7

qixiangyang opened the pull request:
Fix the issue where the prices for most Alibaba Cloud instance types are incorrect when the instance family generation number is greater than or equal to 7.



## What does this PR change?
* According to the customer service and official documentation of Alibaba Cloud, when the instance family generation of an instance is greater than or equal to 7, the system disk does not support "cloud_efficiency". In this case, "cloud_essd" should be used as the default system disk type parameter for queries.

## Does this PR relate to any other PRs?
* No

## How will this PR impact users?
* Get the right price when calculate asset, saving or other result

## Does this PR address any GitHub or Zendesk issues?
* Closes ...

## How was this PR tested?
* add test unit

## Does this PR require changes to documentation?
* No

## Have you labeled this PR and its corresponding Issue as "next release" if it should be part of the next OpenCost release? If not, why not?
* i want label this pr as "next release", just don't know how to do it

Commit:
Fix the issue where the prices for most Alibaba Cloud instance types are incorrect when the instance family generation number is greater than or equal to 7.

Update provider.go

remove debug info

Signed-off-by: qixiangyang <qixiangyangrm@foxmail.com>

teevans commented:
Great catch! We'll have a look and see what we can do!

ameijer commented on pkg/cloud/alibaba/provider.go:
```diff
@@ -1148,6 +1148,24 @@ func getInstanceFamilyFromType(instanceType string) string {
+		}
+	}
+	log.Warnf("unable to find the generation of the instance type %s, returning its default generation as 6", instanceType)
+	return 6
```
is this safe? What specifically about generation 6 makes this suitable? at minimum, please use a const here to avoid the magic number. I would almost return -1 or something so that it clearly is a bogus value

ameijer commented on pkg/cloud/alibaba/provider.go:
```diff
@@ -1148,6 +1148,24 @@ func getInstanceFamilyFromType(instanceType string) string {
+		generation, err := strconv.Atoi(match)
+		if err == nil {
+			return generation
+		}
```
log error

ameijer commented on pkg/cloud/alibaba/provider.go:
```diff
@@ -982,7 +982,7 @@ func createDescribePriceACSRequest(i interface{}) (*requests.CommonRequest, erro
 		} else {
 			// When System Disk information is not available for instance family g6e, r7 and r6e the defaults in
 			// DescribePrice dont default rightly to cloud_essd for these instances.
```
please update this comment to reflect new case

ameijer commented on pkg/cloud/alibaba/provider.go:
```diff
@@ -982,7 +982,7 @@ func createDescribePriceACSRequest(i interface{}) (*requests.CommonRequest, erro
 			// When System Disk information is not available for instance family g6e, r7 and r6e the defaults in
 			// DescribePrice dont default rightly to cloud_essd for these instances.
-			if slices.Contains(alibabaDefaultToCloudEssd, node.InstanceTypeFamily) {
+			if slices.Contains(alibabaDefaultToCloudEssd, node.InstanceTypeFamily) || getInstanceFamilyGenerationFromType(node.InstanceType) > 6 {
```
can you update `alibabaDefaultToCloudEssd` to exclude all >6 generation instance types, so that we don't have multiple conditions that would set this ESSD for gen 7+ instances?

ameijer commented on pkg/cloud/alibaba/provider_test.go:
```diff
@@ -112,6 +112,20 @@ func TestProcessDescribePriceAndCreateAlibabaPricing(t *testing.T) {
+				InstanceTypeFamily: "g8a",
+			},
+			expectedError: nil,
+		},
```
@qixiangyang, I'm concerned that simple checking no error won't ensure that this change works as you expect it to. 

Would you be able to enhance the test case to take in a literal request.CommonRequest that you can then do a deep equal on? Or even just a map of the expected query params. That way, this test would ensure your logic is actually applied in this case

ameijer reviewed (changes requested):
overall looks good, nice catch, a few small requests to improve the testing here and some documentation

ameijer commented:
also, please sign off on your code
![image](https://github.com/user-attachments/assets/f066966f-b6e2-43b4-ba58-4a816b5ed32a)

Commit:
1. Revise 'alibabaDefaultToCloudEssd' exclude 6+ generation instance types.
2. Revise comments when system disk information is not available
3. Add error logging and optimize the return value for the 'getInstanceFamilyGenerationFromType' function.
4. Add test cases for generating node requests for various instance types.

Signed-off-by: xiangyang.qi <xiangyang.qi@advancegroup.com>

1. Revise 'alibabaDefaultToCloudEssd' exclude 6+ generation instance types.
2. Revise comments when system disk information is not available
3. Add error logging and optimize the return value for the 'getInstanceFamilyGenerationFromType' function.
4. Add test cases for generating node requests for various instance types.

Signed-off-by: qixiangyang <qixiangyangrm@foxmail.com>

ameijer commented on pkg/cloud/alibaba/provider.go:
```diff
@@ -1148,6 +1149,26 @@ func getInstanceFamilyFromType(instanceType string) string {
+			return generation
+		}
+	}
+	log.Warnf("unable to find the generation of the instance type %s, returning its default generation as 6", instanceType)
```
```suggestion
	log.Warnf("unable to find the generation of the instance type %s,", instanceType)
```

ameijer commented on pkg/cloud/alibaba/provider.go:
```diff
@@ -49,6 +49,7 @@ const (
 	ALIBABA_UNKNOWN_INSTANCE_FAMILY_TYPE       = "unknown"
 	ALIBABA_NOT_SUPPORTED_INSTANCE_FAMILY_TYPE = "unsupported"
 	ALIBABA_DISK_CLOUD_ESSD_CATEGORY           = "cloud_essd"
+	ALIBABA_DISK_CLOUD_EFFICIENCY_CATEGORY     = "cloud_efficiency"
```
are you using this const anywhere? did you need it for your new test?

qixiangyang replied:
I'm sorry, I forgot to delete it. I did use this variable in the initial test.

qixiangyang committed:
Update pkg/cloud/alibaba/provider.go

Co-authored-by: Alex Meijer <ameijer@users.noreply.github.com>
Signed-off-by: qixiangyang <qixiangyangrm@foxmail.com>

qixiangyang committed:
delete useless const

Signed-off-by: qixiangyang <qixiangyangrm@foxmail.com>

ameijer reviewed (approved):
thank you @qixiangyang !

ameijer committed:
Merge branch 'develop' into develop
